File: aws/CloudFormation/CognitoIdentityProvider/cognitoIdentityProvider.go
package main

import (
	"context"
	"github.com/aws/aws-lambda-go/cfn"
	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/aws/awserr"
	"github.com/aws/aws-sdk-go-v2/aws/external"
	"github.com/aws/aws-sdk-go-v2/service/cognitoidentityprovider"
	"github.com/aws/aws-sdk-go-v2/service/ssm"
	"github.com/codesmith-gmbh/cgc/cgccf"
	"github.com/codesmith-gmbh/forge/aws/common"
	"github.com/mitchellh/mapstructure"
	"github.com/pkg/errors"
	"strings"
)

type Properties struct {
	UserPoolId, ProviderName, ClientIdParameter, ClientSecretParameter string
	ProviderType                                                       cognitoidentityprovider.IdentityProviderTypeType
	AuthorizeScopes                                                    []string
	AttributeMapping                                                   map[string]string
}

func (p *proc) validateProperties(ctx context.Context, input map[string]interface{}) (Properties, map[string]string, error) {
	var properties Properties
	if err := mapstructure.Decode(input, &properties); err != nil {
		return properties, nil, err
	}
	details := make(map[string]string)
	providerType := properties.ProviderType
	switch providerType {
	case cognitoidentityprovider.IdentityProviderTypeTypeGoogle:
		details["authorize_url"] = "https://accounts.google.com/o/oauth2/v2/auth"
		details["authorize_scopes"] = strings.Join(properties.AuthorizeScopes, " ")
		details["attributes_url_add_attributes"] = "true"
		details["token_url"] = "https://www.googleapis.com/oauth2/v4/token"
		details["attributes_url"] = "https://people.googleapis.com/v1/people/me?personFields="
		details["oidc_issuer"] = "https://accounts.google.com"
		details["token_request_method"] = "POST"
	default:
		return properties, nil, errors.Errorf("unknown provider type %s", providerType)
	}
	clientId, err := p.readParameter(ctx, properties.ClientIdParameter)
	if err != nil {
		return properties, nil, err
	}
	clientSecret, err := p.readParameter(ctx, properties.ClientSecretParameter)
	if err != nil {
		return properties, nil, err
	}
	details["client_id"] = clientId
	details["client_secret"] = clientSecret
	return properties, details, nil
}

func main() {
	p := newProc()
	cgccf.StartEventProcessor(p)
}

type proc struct {
	idp *cognitoidentityprovider.Client
	ssm *ssm.Client
}

func newProc() cgccf.EventProcessor {
	cfg, err := external.LoadDefaultAWSConfig()
	if err != nil {
		return &cgccf.ConstantErrorEventProcessor{Error: err}
	}
	return newProcFromConfig(cfg)
}

func newProcFromConfig(cfg aws.Config) *proc {
	return &proc{idp: cognitoidentityprovider.New(cfg), ssm: ssm.New(cfg)}
}

func (p *proc) ProcessEvent(ctx context.Context, event cfn.Event) (string, map[string]interface{}, error) {
	properties, providerDetails, err := p.validateProperties(ctx, event.ResourceProperties)
	if err != nil {
		return "", nil, err
	}
	switch event.RequestType {
	case cfn.RequestDelete:
		return p.deleteIdentityProvider(ctx, event, properties)
	case cfn.RequestUpdate:
		return p.updateIdentityProvider(ctx, event, properties, providerDetails)
	case cfn.RequestCreate:
		return p.createIdentityProvider(ctx, event, properties, providerDetails)
	default:
		return common.UnknownRequestType(event)
	}
}

func (p *proc) readParameter(ctx context.Context, parameterName string) (string, error) {
	decrypt := true
	param, err := p.ssm.GetParameterRequest(&ssm.GetParameterInput{
		Name:           &parameterName,
		WithDecryption: &decrypt,
	}).Send(ctx)
	if err != nil {
		return "", errors.Wrapf(err, "could not read parameter %s", parameterName)
	}
	return *param.Parameter.Value, err
}

func (p *proc) createIdentityProvider(ctx context.Context, event cfn.Event, properties Properties, providerDetails map[string]string) (string, map[string]interface{}, error) {
	_, err := p.idp.CreateIdentityProviderRequest(&cognitoidentityprovider.CreateIdentityProviderInput{
		UserPoolId:       &properties.UserPoolId,
		ProviderName:     &properties.ProviderName,
		ProviderType:     properties.ProviderType,
		AttributeMapping: properties.AttributeMapping,
		ProviderDetails:  providerDetails,
	}).Send(ctx)
	if err != nil {
		return "", nil, err
	}
	data := map[string]interface{}{"UserPoolId": properties.UserPoolId, "ProviderName": properties.ProviderName}
	return properties.UserPoolId + "/" + properties.ProviderName, data, nil
}

func physicalResourceID(properties Properties) string {
	return properties.UserPoolId + "/" + properties.ProviderName
}

func (p *proc) updateIdentityProvider(ctx context.Context, event cfn.Event, properties Properties, providerDetails map[string]string) (string, map[string]interface{}, error) {
	oldUserPoolId := event.OldResourceProperties["UserPoolId"].(string)
	oldProviderName := event.OldResourceProperties["ProviderName"].(string)
	if properties.UserPoolId != oldUserPoolId || properties.ProviderName != oldProviderName {
		return p.createIdentityProvider(ctx, event, properties, providerDetails)
	}
	_, err := p.idp.UpdateIdentityProviderRequest(&cognitoidentityprovider.UpdateIdentityProviderInput{
		UserPoolId:       &properties.UserPoolId,
		ProviderName:     &properties.ProviderName,
		AttributeMapping: properties.AttributeMapping,
		ProviderDetails:  providerDetails,
	}).Send(ctx)
	if err != nil {
		return event.PhysicalResourceID, event.ResourceProperties, errors.Wrapf(err, "could not update the identity provider %s for the user pool %s", properties.ProviderName, properties.UserPoolId)
	}
	return event.PhysicalResourceID, event.ResourceProperties, nil
}

func (p *proc) deleteIdentityProvider(ctx context.Context, event cfn.Event, properties Properties) (string, map[string]interface{}, error) {
	_, err := p.idp.DeleteIdentityProviderRequest(&cognitoidentityprovider.DeleteIdentityProviderInput{
		UserPoolId:   &properties.UserPoolId,
		ProviderName: &properties.ProviderName,
	}).Send(ctx)
	if err != nil {
		awsErr, ok := err.(awserr.RequestFailure)
		if !ok || awsErr.StatusCode() != 400 || awsErr.Code() != "ResourceNotFoundException" {
			return event.PhysicalResourceID, nil, errors.Wrapf(err, "could not delete the identity provider %s", event.PhysicalResourceID)
		}
	}
	return event.PhysicalResourceID, nil, nil
}

File: aws/CloudFormation/CognitoIdentityProvider/cognito_identity_provider.py
import logging

import boto3
from box import Box
from crhelper import CfnResource
from schema import And, Schema

from aws.common.cfn import resource_properties, old_resource_properties, physical_resource_id
from aws.common.schema import not_empty

helper = CfnResource()
logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)

ssm = boto3.client('ssm')
idp = boto3.client('cognito-idp')

properties_schema = Schema({
    'UserPoolId': And(str, not_empty, error='not empty string for UserPoolId'),
    'ProviderName': And(str, not_empty, error='not empty string for ProviderName'),
    'ClientIdParameter': And(str, not_empty, error='not empty string for ClientIdParameter'),
    'ClientSecretParameter': And(str, not_empty, error='not empty string for ClientSecretParameter'),
    'ProviderType': And(str, not_empty, error='not empty string for ProviderType'),
    'AuthorizeScopes': [str],
    'AttributeMapping': {str: str}
})


def validate_properties(properties):
    return Box(properties_schema.validate(properties), camel_killer_box=True)


def derive_details(properties):
    details = {}
    if properties.provider_type == 'Google':
        details["authorize_url"] = "https://accounts.google.com/o/oauth2/v2/auth"
        details["authorize_scopes"] = " ".join(properties.authorize_scopes)
        details["attributes_url_add_attributes"] = "true"
        details["token_url"] = "https://www.googleapis.com/oauth2/v4/token"
        details["attributes_url"] = "https://people.googleapis.com/v1/people/me?personFields="
        details["oidc_issuer"] = "https://accounts.google.com"
        details["token_request_method"] = "POST"
    else:
        raise RuntimeError(f'Unknown provider {properties.provider_type}')
    return details


def read_parameter(parameter_name):
    parameter = ssm.get_parameter(
        Name=parameter_name,
        WithDecryption=True
    )
    return parameter['Parameter']['Value']


@helper.create
def create(event, _):
    properties = validate_properties(resource_properties(event))
    return create_identity_provider(properties)


def create_identity_provider(properties):
    details = derive_details(properties)
    user_pool_id = properties.user_pool_id
    provider_name = properties.provider_name
    idp.create_identity_provider(
        UserPoolId=user_pool_id,
        ProviderName=provider_name,
        ProviderType=properties.provider_type,
        ProviderDetails=details,
        AttributeMapping=properties.attribute_mapping
    )
    helper.Data.update({'UserPoolId': user_pool_id, 'ProviderName': provider_name})
    return f'{user_pool_id}/{provider_name}'


@helper.update
def update(event, _):
    new_properties = validate_properties(resource_properties(event))
    old_properties = validate_properties(old_resource_properties(event))
    if is_new_resource_inferred(new_properties, old_properties):
        return create_identity_provider(new_properties)
    update_identity_provider(new_properties)
    return physical_resource_id(event)


def is_new_resource_inferred(new_properties, old_properties):
    for attr in ['user_pool_id', 'provider_name']:
        if new_properties[attr] != old_properties[attr]:
            return True
    return False


def update_identity_provider(properties):
    details = derive_details(properties)
    idp.update_identity_provider(
        UserPoolId=properties.user_pool_id,
        ProviderName=properties.provider_name,
        ProviderDetails=details,
        AttributeMapping=properties.attribute_mapping
    )


@helper.delete
def delete(event, _):
    properties = validate_properties(resource_properties(event))
    delete_identity_provider(properties)
    return physical_resource_id(event)


def delete_identity_provider(properties):
    try:
        idp.delete_identity_provider(
            UserPoolId=properties.user_pool_id,
            ProviderName=properties.provider_name,
        )
    except idp.exceptions.ResourceNotFoundException:
        pass


def handler(event, context):
    logger.info('event: %s', event)
    helper(event, context)
